internal/appsec/limiter_test.go

```diff
@@ -0,0 +1,251 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2022 Datadog, Inc.
+
+//go:build appsec
+// +build appsec
+
+package appsec
+
+import (
+	"fmt"
+	"math"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestLimiterUnit(t *testing.T) {
+	ticksChan := make(chan time.Time)
+	defer close(ticksChan)
+	ticker := time.Ticker{C: ticksChan}
+	defer ticker.Stop()
+	startTime := time.Now()
+
+	t.Run("no-ticks-1", func(t *testing.T) {
+		l := NewTokenTicker(1, 100)
+		l.start(&ticker, startTime)
+		defer l.Stop()
+		//No ticks between the requests
+		require.True(t, l.Allow(), "First call to limiter.Allow() should return True")
+		require.False(t, l.Allow(), "Second call to limiter.Allow() should return False")
+	})
+
+	t.Run("no-ticks-2", func(t *testing.T) {
+		l := NewTokenTicker(100, 100)
+		l.start(&ticker, startTime)
+		defer l.Stop()
+		//No ticks between the requests
+		for i := 0; i < 10; i++ {
+			require.True(t, l.Allow(), "Call to limiter.Allow() should return True")
+		}
+	})
+
+	t.Run("10ms-ticks", func(t *testing.T) {
+		l := NewTokenTicker(1, 100)
+		l.start(&ticker, startTime)
+		defer l.Stop()
+		require.True(t, l.Allow(), "First call to limiter.Allow() should return True")
+		ticksChan <- startTime.Add(10 * time.Millisecond)
+		require.True(t, l.Allow(), "Second call to limiter.Allow() after 11ms should return True")
+	})
+
+	t.Run("9ms-ticks", func(t *testing.T) {
+		l := NewTokenTicker(1, 100)
+		l.start(&ticker, startTime)
+		defer l.Stop()
+		require.True(t, l.Allow(), "First call to limiter.Allow() should return True")
+		ticksChan <- startTime.Add(9 * time.Millisecond)
+		require.False(t, l.Allow(), "Second call to limiter.Allow() after 9ms should return False")
+	})
+
+	t.Run("1s-rate", func(t *testing.T) {
+		l := NewTokenTicker(1, 1)
+		l.start(&ticker, startTime)
+		defer l.Stop()
+		require.True(t, l.Allow(), "First call to limiter.Allow() should return True with 1s per token")
+		ticksChan <- startTime.Add(500 * time.Millisecond)
+		require.False(t, l.Allow(), "Second call to limiter.Allow() should return False with 1s per Token")
```

Comment:
WDYT of adding one next tick to have Allow returning true?

Reply:
Let's do it